vm: fix race condition with `timeout` param

laverdet opened the pull request:
This fixes a race condition in the watchdog timer used for vm timeouts. The condition would terminate the main stack's execution instead of the code running under the sandbox.

This also fixes the `displayErrors` parameter in the runIn... family of functions.

##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
- vm

##### Notes
The code dealing with `displayErrors` was dead. When this module was first written it looked like this:
```C++
if (try_catch.HasCaught() && try_catch.HasTerminated()) {
  V8::CancelTerminateExecution(args.GetIsolate());
  return ThrowError("Script execution timed out.");
}

if (result.IsEmpty()) {
  // Error occurred during execution of the script.
  DisplayExceptionLine(try_catch.Message());
  try_catch.ReThrow();
  return;
}
```

But 55b87c0238b7cf277d2e872d90bb9729be32ba0e modified the first if in such a way that the second if would never be true. So I went ahead and fixed that while I was in there, but that caused a bunch of errors in the repl tests because the output changed. I disabled `displayErrors` in the repl because the output isn't too useful in that context anyway.

matt-kruse commented:
I just came here looking for someone talking about the problem I'm experiencing. See the code below - timeout exceptions only get caught intermittently. Is this a symptom of the problem you fixed, or is this a different issue?

```
const vm = require('vm');
const timeout = new vm.Script(`while(1){console.log('x');}`);
try {
	timeout.runInNewContext( {"console":console} , {timeout:10} );
}
catch(e) {
	// This should catch a timeout exception, but instead Node just exits.
	// It catches about 5% of the runs successfully.
	// if I remove the try/catch, the exception gets thrown all the way up.
	console.log(e);
}
console.log("done");
```

laverdet commented:
@matt-kruse yes your issue is fixed by this patch :)

matt-kruse commented:
Fantastic insight, thank you. Luckily my actual use case doesn't depend on console or stdout - this was a manufactured test case to play with vm timeouts, and I happened to use console.log().

I'd be curious to know what other landmines might exist in Node when terminating execution abruptly, but I assume as long as my vm code isn't hooking into shared streams or objects, I should be pretty safe.

laverdet commented:
@rodoabad yeah sorry I should have made that clear. `console.error` uses a different WritableStream instance so it still works. You could also still write to stdout using: `  require('fs').write(1, e.stack+ '\n', function() {});`

artch commented:
We use this patch in production in [this project](https://github.com/screeps/screeps). I can confirm that applying it removes a lot of random freezes on `vm` code timeouts.

addaleax commented on lib/repl.js:
```diff
@@ -231,7 +231,7 @@ function REPLServer(prompt,
         try {
           const scriptOptions = {
-            displayErrors: true,
+            displayErrors: false,
```
Any chance you could split the `displayErrors` changes out into a separate commit?

addaleax commented on test/parallel/test-vm-timeout.js:
```diff
@@ -76,3 +76,34 @@ assert.throws(function() {
+// or so. A better, but still incorrect implementation would fail after
+// 100 seconds or so. If you're messing with vm timeouts you might
+// consider increasing this timeout to hammer out races.
+const giveUp = Date.now() + 2500;
```
Tests that are run in our CI shouldn’t generally take more than 1 second to finish… can you split this out into its own test file and reduce the timeout here?

laverdet commented:
Split displayErrors fix out, and added a test for it. I also added a new line to the decorated stack because that matches the behavior of [src/node.cc:1738](https://github.com/nodejs/node/blob/master/src/node.cc#L1738) which is the other place stacks get a fancy arrow added.

Put the vm race test under the `pummel` directory because that seems like the right place?

addaleax commented on src/node_contextify.cc:
```diff
@@ -736,8 +736,10 @@ class ContextifyScript : public BaseObject {
+    Local<String> decorated_stack = String::Concat(
+      String::Concat(arrow.As<String>(),
+        FIXED_ONE_BYTE_STRING(env->isolate(), "\n")),
+      stack.As<String>());
```
Just a tiny stile nit, but we prefer 4 spaces of indentation for statement continuations – that can be fixed while merging, though

addaleax commented on src/node_contextify.cc:
```diff
@@ -956,27 +958,20 @@ class ContextifyScript : public BaseObject {
     bool timed_out = false;
     bool received_signal = false;
```
I think in a previous version of this PR these were `volatile`… any reason why that changed?

laverdet replied:
Yeah sorry I meant to leave a note about that. The volatile ended up being unnecessary. The watchdogs take a non-const reference so that's good enough to ensure the optimizer doesn't do anything fishy. Then in the dtor `uv_thread_join` is called which will synchronize the watchdog thread back up with the main thread.

addaleax commented on src/node_watchdog.h:
```diff
@@ -37,16 +37,14 @@ namespace node {
+  explicit Watchdog(
+    v8::Isolate* isolate,
+    uint64_t ms,
+    bool* timed_out = NULL);
```
Ditto for 4-space indents. I think here we’d usually keep the `v8::Isolate* isolate` bit directly after the `(` and align all other parameters vertically below it; but again, this is just a tiny nit and can be fixed up while landing.

addaleax commented on src/node_watchdog.h:
```diff
@@ -56,27 +54,21 @@ class Watchdog {
-  explicit SigintWatchdog(v8::Isolate* isolate);
+  explicit SigintWatchdog(
+    v8::Isolate* isolate,
+    bool* received_signal = NULL);
```
(ditto)

addaleax commented on test/pummel/test-vm-race.js:
```diff
@@ -0,0 +1,54 @@
+// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+// USE OR OTHER DEALINGS IN THE SOFTWARE.
```
You can drop the copyright header for newly added files, I think.

laverdet commented:
With regard to the style changes is there a style guide out there for newcomers? The jslinter included was pretty helpful but in C++ land I found a lot of clashing styles so I just winged it. Though now that you point it out the 4-space continuation thing seems obvious.

addaleax commented:
> With regard to the style changes is there a style guide out there for newcomers?

Nothing except the [Google styleguide](https://google.github.io/styleguide/cppguide.html) which we roughly follow, but tbh I wouldn’t expect anybody to read the full text for making a contribution (e.g. I haven’t done that myself). We could probably tighten up the linter, but it doesn’t usually take people long to get used to the C++ style quirks that we have; I think the 4-space indents for continuations, `lower_snake_case` for local variables and preferring to align arguments vertically are the only things that trip up people on a regular basis.

> Though now that you point it out the 4-space continuation thing seems obvious.

It took me a while to get used to it as well ;)

addaleax commented:
Landed in 3c5bfba28bc9ed244daed7e8f8234b9a4529be85 and e018c331114c5a8e73e374cb493730e7dcc9bc91, fixed up the left nits while landing. Thank you a lot for the PR!

MylesBorins commented:
Should this land on v6.x?

laverdet commented:
@MylesBorins the patches apply cleanly to the v6.11.1 tag for me, except for the output of a test which is a simple fix. Did you run into other issues?

MylesBorins commented:
@laverdet that seems to be the conflict. Unfortunately the fix is not so simple for me to sort out... could you provide a Backport PR please

laverdet commented:
@MylesBorins please see #14373 for the backport :)
